src/org/apache/xpath/functions/FuncFilter.java

```diff
@@ -0,0 +1,237 @@
+           return resultSeq;
+        }
+        
+        QName varQname = null;
+        
+        if (funcParamNameList.size() == 1) {
+           varQname = new QName(funcParamNameList.get(0));
+        }
+        
+        SourceLocator srcLocator = xctxt.getSAXLocator();
+        
+        XPath xpathInlineFn = new XPath(funcBodyXPathExprStr, srcLocator, null, XPath.SELECT, null);
+        
+        if (arg0XsObject instanceof ResultSequence) {
+           XPathContext xpathContextNew = new XPathContext(false);
+           Map<QName, XObject> inlineFunctionVarMap = xpathContextNew.getInlineFunctionVarMap();
+        
+           ResultSequence inpResultSeq = (ResultSequence)arg0XsObject;
+           for (int idx = 0; idx < inpResultSeq.size(); idx++) {
+               XObject inpSeqItem = inpResultSeq.item(idx);
+               if (varQname != null) {
+                  inlineFunctionVarMap.put(varQname, inpSeqItem);
+               }
+        
+               XObject resultObj = xpathInlineFn.execute(xpathContextNew, DTM.NULL, null);
+               if (resultObj instanceof XBoolean) {
+                   if (((XBoolean)resultObj).bool()) {
+                      resultSeq.add(inpSeqItem);
+                   }
+               }
+               else {
+                   throw new javax.xml.transform.TransformerException("XPTY0004 : The item type of the result of calling "
+                                                                                              + "function filter() is not xs:boolean.", xctxt.getSAXLocator()); 
+               }
+           }
+        
+           inlineFunctionVarMap.clear();
+        }
+        else if (arg0DtmIterator != null) {                  
+           Map<QName, XObject> inlineFunctionVarMap = xctxt.getInlineFunctionVarMap();
+        
+           int dtmNodeHandle;
+           
+           while (DTM.NULL != (dtmNodeHandle = arg0DtmIterator.nextNode())) {
+               DTM dtm = xctxt.getDTM(dtmNodeHandle);
+               Node node = dtm.getNode(dtmNodeHandle);
+               XObject inpSeqItem = XObjectFactory.create(node, xctxt);               
+               if (varQname != null) {
+                  inlineFunctionVarMap.put(varQname, inpSeqItem);
+               }
+        
+               xctxt.pushCurrentNode(dtmNodeHandle);
+               
+               XObject resultObj = xpathInlineFn.execute(xctxt, dtmNodeHandle, null);
+               if (resultObj instanceof XBoolean) {
+                   if (((XBoolean)resultObj).bool()) {                       
+                       resultSeq.add(new XNodeSet(dtmNodeHandle, xctxt));
+                   }
+               }
+               else {
+                   throw new javax.xml.transform.TransformerException("XPTY0004 : The item type of the result of calling "
+                                                                                              + "function filter() is not xs:boolean.", xctxt.getSAXLocator());  
```

Comment:
It would help if the error message could refer to the type (and value) of the actual resut